### PR TITLE
Disable compiled autograd for re-entrant autograd

### LINKED ISSUE #135298: compiled_autograd deadlock when using checkpoint
### 🐛 Describe the bug

when we try to use compiled_autograd to improve the backward performance ( graph break makes checkpoint fallback to eager, all fwd and bwd run in eager got bad performance)
when we call `out.sum().backward()`,  the  `CheckpointFunction.backward` will call `torch.autograd.backward`, then leads to a deadlocked.

I have a workaround by now, I'm not sure if it can be taken as a final solution. 
another option is disable it in `torch::dynamo::autograd::compiled_autograd(...)` before lock.

```python
with torch._dynamo.compiled_autograd.disable():
    torch.autograd.backward(...)

```

### Error logs

_No response_

### Minified repro

```python
import os
import sys
import numpy as np

os.environ["TORCH_LOGS"] = "dynamo,graph_breaks,recompiles,graph_code,aot_joint_graph,aot_graphs"
os.environ["TORCHDYNAMO_VERBOSE"] = "1"
os.environ["TORCHINDUCTOR_COMPILE_THREADS"] = "1"
import torch
import torch._dynamo

os.environ["PT_HPU_LAZY_MODE"] = "0"
os.environ["ENABLE_CONSOLE"] = "1"
os.environ["LOG_LEVEL_ALL"] = "6"
os.environ["LOG_LEVEL_PT_EAGER"] = "1"

class MyModule(torch.nn.Module):
    def __init__(self, device):
        super(MyModule, self).__init__()
        self.layer = MyModule2(device)

    def forward(self, x):
        torch._dynamo.graph_break()
        res = x.cos() - x.sin()
        return self.layer(res)

class MyModule2(torch.nn.Module):
    def __init__(self, device):
        super(MyModule2, self).__init__()
        self.linear = torch.nn.Linear(2, 2, device=device)

    def forward(self, x):
        return self.linear(x).cos()


# stock pytorch checkpoint
def stock(args):
    device = args[0] if args else "cpu"

    from torch.utils.checkpoint import checkpoint
    
    if device == "cpu":
        def enable_compiled_autograd():
            def compiler_fn(gm):
                #return torch.compile(gm, backend='inductor', fullgraph=False)
                return torch.compile(gm, backend='eager', fullgraph=False)

            import functools
            import torch
            from torch._dynamo import compiled_autograd
            torch._C._dynamo.compiled_autograd.set_autograd_compiler(
                    functools.partial(compiled_autograd.AutogradCompilerInstance, compiler_fn)
                )

            torch._dynamo.reset()
            torch._dynamo.config.optimize_ddp = "python_reducer"
    else:
        import habana_frameworks.torch.dynamo.compile_backend
        from habana_frameworks.torch.dynamo.compile_backend.experimental import enable_compiled_autograd

    enable_compiled_autograd()

    m = MyModule(device)
    def fn(x):
        x = checkpoint(m.__call__, x)
        out = x + 2 
        out = out * 2 
        return out


    # fn = torch.compile(fn, backend="inductor" if device == "cpu" else "hpu_backend")
    fn = torch.compile(fn, backend="hpu_backend" if device == "hpu" else "inductor")
    for i in range(2):
        print(f"run {i} ..., ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff")
        x = torch.ones(2, 2, device=device, requires_grad=True)
        out = fn(x)
        print(out.cpu())
        print(f"backward {i} ..., bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb")
        out.sum().backward()


if __name__ == "__main__":

    seed = 1000
    torch.manual_seed(seed)
    # habana_frameworks.torch.hpu.random.manual_seed(seed)

    if len(sys.argv) > 1:
        try:
            eval(f"{sys.argv[1]}({sys.argv[2:]})")
        except Exception as e:
            print(e)
            l = []
            for key, value in dict(locals()).items():
                if callable(value) and value.__module__ == __name__:
                    l.append(key)
            print(f"you can run one of the following functions: {l}")
        exit(0)
```

### Versions

 python pytorch-fork/torch/utils/collect_env.py
Collecting environment information...
PyTorch version: 2.4.0a0+git12138a8
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.5 (ssh://gerrit.habana-labs.com:29418/tpc_llvm10 0c338a08ec38e26cef18216f902a07849a108a56)
CMake version: version 3.28.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jul 29 2024, 16:56:48) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-117-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        43 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               12
On-line CPU(s) list:                  0-11
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Gold 6132 CPU @ 2.60GHz
CPU family:                           6
Model:                                85
Thread(s) per core:                   1
Core(s) per socket:                   6
Socket(s):                            2
Stepping:                             0
BogoMIPS:                             5187.81
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon nopl xtopology tsc_reliable nonstop_tsc cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 invpcid avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xsaves arat pku ospke md_clear flush_l1d arch_capabilities
Virtualization:                       VT-x
Hypervisor vendor:                    VMware
Virtualization type:                  full
L1d cache:                            384 KiB (12 instances)
L1i cache:                            384 KiB (12 instances)
L2 cache:                             12 MiB (12 instances)
L3 cache:                             38.5 MiB (2 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-11
Vulnerability Gather data sampling:   Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:          KVM: Mitigation: VMX disabled
Vulnerability L1tf:                   Mitigation; PTE Inversion; VMX flush not necessary, SMT disabled
Vulnerability Mds:                    Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:               Mitigation; PTI
Vulnerability Mmio stale data:        Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; IBRS; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] flake8==7.0.0
[pip3] habana-torch-dataloader==1.18.0+git0092472b4
[pip3] habana-torch-plugin==1.18.0+git0092472b4
[pip3] mypy==1.10.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] pylsp-mypy==0.6.8
[pip3] pytorch-lightning==2.2.0.post0
[pip3] torch==2.4.0a0+git12138a8
[pip3] torch-debug==2.2.0a0+git10d65c0
[pip3] torch_tb_profiler==0.4.0
[pip3] torchaudio==2.4.0a0+69d4077
[pip3] torchdata==0.7.1+5e6f7b7
[pip3] torchmetrics==1.3.2
[pip3] torchtext==0.18.0a0+9bed85d
[pip3] torchvision==0.19.0a0+48b1edf
[pip3] triton==2.2.0
[conda] Could not collect

cc @ezyang @chauhang @penguinwu @xmfan

### PR DESCRIPTION
**Pull request by majian4work**
Fixes #135298

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

### CODE REVIEW

**Comment by majian4work** [comment]
@xmfan please help to take a review

File: torch/autograd/graph.py [hunk 1]
```diff
@@ -821,10 +821,16 @@ def _engine_run_backward(
...
+    compiled_autograd_ctx = (
+        torch._dynamo.compiled_autograd.disable()
+        if torch._dynamo.compiled_autograd.in_compiled_autograd_region
+        else contextlib.nullcontext()
+    )
     try:
-        return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
-            t_outputs, *args, **kwargs
-        )  # Calls into the C++ engine to run the backward pass
+        with compiled_autograd_ctx:
```
Comment (xmfan) [inline torch/autograd/graph.py:830]: could we move the context creation logic into some helper in torch/_dynamo/compiled_autograd.py? and keep minimal logic here

```suggestion
        with torch._dynamo.compiled_autograd.maybe_disable_compiled_autograd():
```

Comment (majian4work) [reply torch/autograd/graph.py]: done

**Comment by majian4work** [comment]
@xmfan could you help to retry the failed test?

**Review by xmfan — DISMISSED** [review]

**Comment by xmfan** [comment]
A small number of internal tests that were timing the backward's wall time started reporting increased latency. The only thing I can think of is the nullcontext entry.

Could we branch the logic with an if else such that non-compiled autograd is minimally affected?

**Comment by majian4work** [comment]
hi @xmfan 
It's easy to update this patch using the branch, but how can I verify the performance result?

**Comment by xmfan** [comment]
I'm importing this PR into our internal codebase to run the checks again

**Comment by xmfan** [comment]
@majian4work  So the reason why internal checks are failing is because of high overhead from the `torch._dynamo.compiled_autograd.in_compiled_autograd_region` access. So we can't add it to the hotpath. Another way to always disable compiled autograd at compiled autograd runtime. I can't modify the PR, so here's a patch that you can apply with `git apply <file>`: https://gist.github.com/xmfan/1899e87cfc4967c580fe303bbccd0833

**Review by xmfan — APPROVED** [review]

**Comment by majian4work** [comment]
weirdly, the UT failure can't be reproduced locally.

**Pull request closed without merging**